sonar-batch/src/main/java/org/sonar/batch/bootstrap/WSLoader.java

```
@@ -0,0 +1,182 @@
+
+  public enum ServerStatus {
+    UNKNOWN, ACCESSIBLE, NOT_ACCESSIBLE;
+  }
+
+  public enum LoadStrategy {
+    SERVER_FIRST, CACHE_FIRST;
+  }
+
+  private LoadStrategy loadStrategy;
+  private boolean cacheEnabled;
+  private ServerStatus serverStatus;
+  private ServerClient client;
+  private PersistentCache cache;
+
+  public WSLoader(boolean cacheEnabled, PersistentCache cache, ServerClient client) {
+    this.cacheEnabled = cacheEnabled;
+    this.loadStrategy = CACHE_FIRST;
+    this.serverStatus = UNKNOWN;
+    this.cache = cache;
+    this.client = client;
+  }
+
+  public WSLoader(PersistentCache cache, ServerClient client) {
+    this(false, cache, client);
+  }
+
+  public ByteSource loadSource(String id) {
+    return ByteSource.wrap(load(id));
+  }
+
+  public String loadString(String id) {
+    return new String(load(id), StandardCharsets.UTF_8);
+  }
+
+  @Nonnull
+  public byte[] load(String id) {
+    if (loadStrategy == CACHE_FIRST) {
+      return loadFromCacheFirst(id);
+    } else {
+      return loadFromServerFirst(id);
+    }
+  }
+
+  public void setStrategy(LoadStrategy strategy) {
+    this.loadStrategy = strategy;
+  }
+
+  public LoadStrategy getStrategy() {
+    return this.loadStrategy;
+  }
+
+  public void setCacheEnabled(boolean enabled) {
+    this.cacheEnabled = enabled;
+  }
+
+  public boolean isCacheEnabled() {
+    return this.cacheEnabled;
+  }
+
+  private void switchToOffline() {
+    serverStatus = NOT_ACCESSIBLE;
+  }
+
+  private void switchToOnline() {
+    serverStatus = ACCESSIBLE;
+  }
+
+  private boolean isOffline() {
+    return serverStatus == NOT_ACCESSIBLE;
+  }
+
+  @Nonnull
+  private byte[] loadFromCacheFirst(String id) {
+    byte[] cached = loadFromCache(id);
+    if (cached != null) {
+      return cached;
+    }
+
+    try {
+      return loadFromServer(id);
+    } catch (Exception e) {
+      if (e.getCause() instanceof HttpDownloader.HttpException) {
```

Is it intended to silently ignore all other exceptions?